upstairs/src/upstairs.rs

```diff
@@ -0,0 +1,3370 @@
+             */
+            if !matches!(self.state, UpstairsState::GoActive { .. }) {
+                info!(
+                    self.log,
+                    "could not connect region set due to bad state: {:?}",
+                    self.state
+                );
+                return false;
+            }
+            /*
+             * Make sure all downstairs are in the correct state before we
+             * proceed.
+             */
+            let not_ready = self
+                .downstairs
+                .clients
+                .iter()
+                .any(|c| c.state() != DsState::WaitQuorum);
+            if not_ready {
+                info!(
+                    self.log,
+                    "Waiting for {} more clients to be ready", not_ready
+                );
+                return false;
+            }
+
+            /*
+             * We figure out if there is any reconciliation to do, and if so, we
+             * build the list of operations that will repair the extents that
+             * are not in sync.
+             *
+             * If we fail to collate, then we need to kick out all the
+             * downstairs out, forget any activation requests, and the
+             * upstairs goes back to waiting for another activation request.
+             */
+            self.downstairs.collate(self.generation)
+        };
+
+        match collate_status {
+            Err(e) => {
+                error!(self.log, "Failed downstairs collate with: {}", e);
+                // We failed to collate the three downstairs, so we need
+                // to reset that activation request.  Call
+                // `abort_reconciliation` to abort reconciliation for all
+                // clients.
+                self.set_inactive(e);
+                self.downstairs.abort_reconciliation(&self.state);
+                false
+            }
+            Ok(true) => {
+                // We have populated all of the reconciliation requests in
+                // `Downstairs::reconcile_task_list`.  Start reconciliation by
+                // sending the first request.
+                self.downstairs.send_next_reconciliation_req().await;
+                true
+            }
+            Ok(false) => {
+                info!(self.log, "No downstairs repair required");
+                self.on_reconciliation_done(DsState::WaitQuorum);
+                info!(self.log, "Set Active after no repair");
+                true
+            }
+        }
+    }
+
+    /// Called when reconciliation is complete
+    fn on_reconciliation_done(&mut self, from_state: DsState) {
+        // This should only ever be called if reconciliation completed
+        // successfully; make some assertions to that effect.
+        self.downstairs.on_reconciliation_done(from_state);
+
+        info!(self.log, "All required repair work is completed");
+        info!(self.log, "Set Downstairs and Upstairs active after repairs");
+
```

Comment:
I can't convince myself that it's impossible – a deactivation request from the guest wouldn't do it, because we'd reject it [here](https://github.com/oxidecomputer/crucible/blob/bfc76dd9e28715f5cc22157de4b1c0277b00aead/upstairs/src/upstairs.rs#L988), but a call to `set_inactive` could do it, and it's called from a bunch of places...

Reply:
I added some additional checking in 317ede3cb2c775a86b88fe36dc93f2e1a2c9440d, so that this logs an error but does not panic on a failure.